### PR TITLE
feat: FlexibleProcessing as alternative to EntityComponentProcessor for lightweight, abstract and typesafe processing in idiomatic c#

### PR DESCRIPTION
**Pull request by Eideren**
# PR Details
This PR enables building of processors from interfaces, enabling unrelated types to be processed by the same system. 

The current processor implementation force users to derive from the same root type to build their logic, this does not work well with the fact that we have different 'Script' variant which have all their specific use. It becomes a composition issue when more abstract concepts, like event functions, are introduced. 

A good example of that is when components should manipulate physics object, it is preferable to do so in lockstep with the physics engine's tick. But those component may also require an async or update loop to process inputs. We would be forced to either create another new 'script' type which users have to derive from which would split their logic between two different types, or tie it to one of the other 'script', both solutions would introduce additional overhead while not being practical.

Example usage:
```cs
public class MyComponent : StartupScript, IPhysicsTick
{
    public void Tick(float deltaTime){ }
}

public interface IPhysicsTick : IComponent<IPhysicsTick.PhysicsTickProc, IPhysicsTick>
{
    void Tick(float deltaTime);
    public class PhysicsTickProc : IProcessor, IUpdateProcessor
    {
        public int Order => 5;
        public List<IPhysicsTick> Components = new();
        public void SystemAdded(IServiceRegistry registryParam) { }
        public void SystemRemoved() { }
        public void OnComponentAdded(IPhysicsTick item) => Components.Add(item);
        public void OnComponentRemoved(IPhysicsTick item) => Components.Remove(item);
        public void Update(GameTime gameTime)
        {
            foreach (var comp in Components)
                comp.Tick((float)gameTime.Elapsed.TotalSeconds);
        }
    }
}

```
Implementer may include additional processing during the `OnComponentAdded` step to reduce or elide virtual calls

## Related Issue
Introduced for usage in #2131 

## Types of changes
- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [ ] My change requires a change to the documentation.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] **I have built and run the editor to try this change out.**

### CODE REVIEW

File: sources/engine/Stride.Engine/Engine/FlexibleProcessing/ProcessorManager.cs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        }
+
+        public void RemoveComponent(EntityComponent _component)
+        {
+            if (_component is IMarkedComponent component == false)
+                return;
+
+            foreach (var procIndex in compTypeToProcessorIndices[component.GetType()])
+            {
+                var pData = processors[procIndex];
```
Comment (manio143) [inline sources/engine/Stride.Engine/Engine/FlexibleProcessing/ProcessorManager.cs:109]: Would accessing the list as Span allow an inline modification here?
`processors.AsSpan()[procIndex].ComponentCount--`

**Review by manio143 — COMMENTED** [review]
Nice work!

File: sources/engine/Stride.Engine/Engine/FlexibleProcessing/ProcessorManager.cs (hunk 1 above)
Comment (Eideren) [reply sources/engine/Stride.Engine/Engine/FlexibleProcessing/ProcessorManager.cs]: Yep, I thought of using that initially but at the time the logic was set up in a different way which benefited from copying the data, now it fits in like a glove, thanks !

File: sources/engine/Stride.Engine/Engine/FlexibleProcessing/ProcessorManager.cs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                {
+                    pData.Instance = (IProcessorBase)Activator.CreateInstance(pData.Type)!;
+                    if (pData.Instance is IUpdateProcessor update)
+                        processorsToUpdate.Add(new(update.Order, pData.Type), new(update, new ProfilingKey(GameProfilingKeys.GameUpdate, GetType().Name)));
+                    if (pData.Instance is IDrawProcessor draw)
+                        processorsToDraw.Add(new(draw.Order, pData.Type), new(draw, new ProfilingKey(GameProfilingKeys.GameDraw, GetType().Name)));
+
+                    processors[procIndex] = pData; // Set beforehand just in case SystemAdded adds a component
+                    pData.Instance!.SystemAdded(registry);
+                    pData = processors[procIndex]; // Refresh local for the same reason
```
[the remark this answers is not in the document]
Comment (Eideren) [reply sources/engine/Stride.Engine/Engine/FlexibleProcessing/ProcessorManager.cs]: Yep, that's it. Using a class would have been more straightforward but it wasn't that much more complex to use a struct and would improve cache throughput.

File: sources/engine/Stride.Engine/Engine/EntityManager.cs [hunk 3]
```diff
@@ -492,6 +503,11 @@ private void CheckEntityWithProcessors(Entity entity, bool forceRemove, bool col
                 {
                     CollectNewProcessorsByComponentType(component.GetType().GetTypeInfo());
                 }
+
+                if (forceRemove)
+                    flexibleProcessors.RemoveComponent(component);
```
[the remark this answers is not in the document]
Comment (Eideren) [reply sources/engine/Stride.Engine/Engine/EntityManager.cs:508]: Yeah, the entity manager's logic is ridiculously obtuse for how simple the operations actually are.
This function is only ever called when an entity is added or removed from the scene (`forceRemove` specifies whether the entity is added when false, and removed when true), this function is never called when removing a specific component of an entity that's already part of the scene/manager.
Basically, all of the entity's components should be removed from all processors when `forceRemove` is `true`.

[the remark this answers is not in the document]
Comment (Eideren) [reply sources/engine/Stride.Engine/Engine/EntityManager.cs:508]: Still, I added a test validating this claim.

**Review by manio143 — APPROVED** [review]

**Comment by Eideren** [comment]
Thanks a bunch for reviewing this @manio143

**Pull request merged**